analyzer/src/main/kotlin/managers/utils/DependencyGraphBuilder.kt

```diff
@@ -0,0 +1,324 @@
+ * packages depend. A naive implementation, which simply duplicates the dependency trees for each scope and package
+ * therefore leads to a high consumption of memory. This class addresses this problem by generating an optimized
+ * structure that shares the components of the dependency graph between the scopes as far as possible.
+ *
+ * This builder class provides the _addDependency()_ function, which has to be called for all the direct dependencies
+ * of the different scopes. From these dependencies it constructs a single, optimized graph that is referenced by all
+ * scopes. To reduce the amount of memory required even further, package identifiers are replaced by numeric indices,
+ * so that references in the graph are just numbers.
+ *
+ * Ideally, the resulting dependency graph contains each dependency exactly once. There are, however, cases, in which
+ * packages occur multiple times in the project's dependency graph with different dependencies. Such cases are
+ * detected, and the corresponding packages form different nodes in the graph, so that they can be distinguished
+ * correctly.
+ *
+ * This class implements the full logic to construct a [DependencyGraph], independent on the concrete representation of
+ * dependencies used by specific package managers. To make this class compatible with such a dependency representation,
+ * the package manager implementation has to provide a [DependencyHandler]. Via this handler, all the relevant
+ * information about dependencies can be extracted.
+ */
+class DependencyGraphBuilder<D>(
+    /**
+     * The [DependencyHandler] used by this builder instance to extract information from the dependency objects when
+     * constructing the [DependencyGraph].
+     */
+    val dependencyHandler: DependencyHandler<D>
+) {
+    companion object {
+        /**
+         * Internal class to represent the result of a search in the dependency graph. The outcome of the search
+         * determines how to integrate a specific dependency into the dependency graph.
+         */
+        private sealed class GraphSearchResult {
+
+            /**
+             * A specialized [GraphSearchResult] that indicates that the dependency that was searched for is already
+             * present in the dependency graph. This is the easiest case, as the [DependencyReference] that was found
+             * can directly be reused.
+             */
+            data class Found(
+                /** The reference to the dependency that was searched in the graph. */
+                val ref: DependencyReference
+            ) : GraphSearchResult()
+
+            /**
+             * A specialized [GraphSearchResult] that indicates that the dependency that was searched for was not found
+             * in the dependency graph, but a fragment has been detected, to which it can be added.
+             */
+            data class NotFound(
+                /** The index of the fragment to which to add the dependency. */
```

Comment:
Can you explain what a fragment is somewhere? I found it a bit difficult to figure out the idea behind it.

Reply:
This is indeed a term that is not easy to grasp and explain. I have tried to make it more clearer in the class comment of `DependencyGraphBuilder`.